src/init.c

```diff
@@ -93,6 +95,8 @@ void __attribute__ ((constructor)) icainit(void)
 		ica_set_stats_mode(value);
 
 #ifdef ICA_FIPS
+	static CRYPTO_ONCE once_run = CRYPTO_ONCE_STATIC_INIT;
+	CRYPTO_THREAD_run_once(&once_run, fips_lib_integrity_check);
```

Comment:
we are in the constructor here. any reason to use libcrypto's run_once ? and why did you put it before fips_init ? Did you check if fips imposes an order on the start up checks ?

Reply:
I think the integrity check is part of the power up test and should be called from fips_powerup_tests (fips.c). If it fails it should set the ICA_FIPS_INTEGRITY (ica_api.h) in the global fips module status flags (fips variable in fips.c). That is how it was supposed to work.

Reply:
ok, you're right,  the run_once is superfluous. 
Re: order on the fips start up checks: I don't think there is an order. If running in fips mode and any of the tests fails, the lib should terminate. But we may move the integrity check after the power up test. 
Re: ICA_FIPS_INTEGRITY: I wasn't aware of this flag. OK, I will handle this also in fips_test.c similar to ICA_FIPS_CRYPTOALG.
But what about the exit(1); in fips_lib_integrity_check? I think the lib must exit if this check fails.

Reply:
See FIPS 140-2 chap. 4.9: 

Self-tests:
- power-up test (algorithms, integrity, critical functions)
- conditional tests

So the integrity test is part of the power up tests (and thus should be called from libica's fips_powerup_tests). The spec doesnt impose an order of the power-up tests, but id just use the order in which the specs lists the tests i.e., but it below the cryptographic algorithm tests in the power up test function. (the power up test's critical function test are n/a for libica).

Regarding exit: the spec says:
```
If a cryptographic module fails a self-test, the module shall enter an error state and output an error indicator via the status output interface.  The cryptographic module shall not perform any cryptographic operationswhile  in  an  error  state.  All  data  output  via  the  data  output  interface  shall  be  inhibited  when  an  error  state  exists. 
```

So right now, in case of fips error state, all libica functions will return a corresponding error. Do you think this is not enough? Why do you think the library, resp. the application which links the library, should always exit ?

Reply:
Hm, not sure if a library should ever do exit(). This could end a process silently, and nobody knows why. So if you need to exit the hard way, then use abort(), this will at least be seen by the user.
I am still not sure if a library should abort, or influence the life of the calling process in any way. Maybe that's what we want for the FIPS checks though....

Reply:
OK, doing the exit() was the natural way for me to handle this. But when the spec says we should go into an error state this is also fine. It now just sets the ICA_FIPS_INTEGRITY flag. I also moved the integrity check to the power up test.

Reply:
One last question: Do we want an additional fprintf when the check fails?
```
	if (!FIPSCHECK_verify(path)) {
		syslog(LOG_ERR, msg2, path);
		fprintf(stderr, msg2, path);
		fips |= ICA_FIPS_INTEGRITY;
	}
```
When the check fails, we get a 
> FIPS SELF-TEST FAILURE. CHECK THE SYSLOG.

So an additional message like
> Libica FIPS library integrity check failed. Module /usr/local/lib/libica.so.3 probably corrupted.

is not really necessary. Other fips tests don't issue extra messages.
What do you think?

Reply:
> Do we want an additional fprintf when the check fails?

No, the library should not print to stderr/out.

> FIPS SELF-TEST FAILURE. CHECK THE SYSLOG.

This message is displayed by the icainfo tool, not printed by the library on failure.

A syslog entry for the integrity check failure is enough, in case other fips sleft tests fail, we do the same e.g.:
> syslog(LOG_ERR, "Libica internal RNG error..");